runtime/near-vm-runner/RUNTIMES.md

```diff
@@ -0,0 +1,214 @@
+the runtime execute the instructions (throughput), execute the first instruction (latency), prepare
+the initial code for execution (compilation) and how efficient are miscellaneous operations such as
+serialization and deserialization of the module.
+
+In order to realistically evaluate the performance of a runtime it is important to understand how
+the NEAR validator utilizes the runtime to implement its functionality. Somewhat simplified view
+is:
+
+1. A module containing Wasm code in binary format is parsed and instrumented;
+2. Wasm code produced after instrumentation is supplied to the runtime for preparation
+   (compilation);
+3. The prepared module is serialized and written to a long term storage (database);
+4. When the contract function call occurs, the serialized module is deserialized, loaded into
+   memory and the execution of the desired function begins (latency);
+5. The contract code is executed (throughput) and the results are returned to the validator.
+
+It is important to remember that the NEAR protocol charges fees for an operation before the
+operation is executed. For that reason, predictability of worst case execution time often matters
+more than the execution time in the typical case. It is important that we are able to deduce ahead
+of time what cost to assign to a given operation. Inability to do so can lead to significant
+undercharging and break the properties underlying the protocol.
+
+On a scale of trade-offs, performance is probably one of the less important metrics. Slower
+execution of a contract code doesn't make NEAR protocol unsound as an idea and only serves to
+impose stricter limits on what can be achieved on the network.
+
+### Compilation
+
+The NEAR protocol charges gas fees to deploy a contract. One part of a deployment involves
+preparing the contract Wasm code for execution. This may involve instrumentation, compilation,
+serialization of the machine code and other similar operations.
+
+An estimate of the deployment cost will typically involve a function which uses the size of the
+input Wasm code as its primary input. Such a function can only exist if we have a good knowledge of
+the runtime’s time complexity properties are known. For our purposes a linear or `O(n log n)`
+relationship between the input size and execution time is the highest we can accept.
+
+### Latency
+
+Whenever a function within a contract is invoked, the serialized form of the contract will be
+instantiated: loaded into memory and prepared for execution (though processes such as e.g.
+linking). Only then it is possible to start executing the contract code.
+
+When executing a tiny function part of a larger contract these operations will dominate and
+contribute greatly to the observed latency of the contract execution. These overheads contribute to
+the fees paid by anybody using the protocol, making any unnecessary overhead a potential roadblock
+in NEAR protocol's adoption.
```

Comment:
N00b question: is the overhead here fundamentally unavoidable, i.e, is there some way to only deserialize part of the compiled wasm module to reduce the cost when only a tiny function is invoked?

Reply:
In theory it is possible to only load the functions necessary for execution by constructing a call graph. In practice a non-negligible number of contracts will utilize the `call_indirect` instruction which makes this sort of analysis much more complicated.

An alternative could be to load the machine code into memory lazily, as it is accessed. This is exactly what operating systems like Linux do when they execute a program.